src/net_processing.cpp

```diff
@@ -4365,6 +4368,68 @@ void PeerManagerImpl::MaybeSendPing(CNode& node_to, Peer& peer)
     }
 }
 
+void PeerManagerImpl::MaybeSendAddr(CNode& pto)
+{
+    // Nothing to do for non-address-relay peers
+    if (!pto.RelayAddrsWithConn()) return;
+
+    auto current_time = GetTime<std::chrono::microseconds>();
+
+    // Periodically advertise our local address to the peer.
+    if (fListen && !::ChainstateActive().IsInitialBlockDownload() &&
+        pto.m_next_local_addr_send < current_time) {
+        // If we've sent before, clear the bloom filter for the peer, so that our
+        // self-announcement will actually go out.
+        // This might be unnecessary if the bloom filter has already rolled
+        // over since our last self-announcement, but there is only a small
+        // bandwidth cost that we can incur by doing this (which happens
+        // once a day on average).
+        if (pto.m_next_local_addr_send != 0us) {
+            pto.m_addr_known->reset();
+        }
+        if (Optional<CAddress> local_addr = GetLocalAddrForPeer(&pto)) {
+            FastRandomContext insecure_rand;
+            pto.PushAddress(*local_addr, insecure_rand);
+        }
+        pto.m_next_local_addr_send = PoissonNextSend(current_time, AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL);
+    }
+
+    // We sent an `addr` message to this peer recently. Nothing more to do.
+    if (current_time <= pto.m_next_addr_send) return;
+    pto.m_next_addr_send = PoissonNextSend(current_time, AVG_ADDRESS_BROADCAST_INTERVAL);
+
+    assert(pto.m_addr_known);
+    assert(pto.vAddrToSend.size() <= MAX_ADDR_TO_SEND);
```

Comment:
It's currently impossible for `vAddrToSend` to grow larger than `MAX_ADDR_TO_SEND` (observe that the only place where elements are added to `vAddrToSend` check the size first). However, it might be safer to do something like:

```suggestion
    if(pto.vAddrToSend.size() > MAX_ADDR_TO_SEND) vAddrToSend.resize(MAX_ADDR_TO_SEND);
```

I'm happy to go with whatever reviewers prefer.

Reply:
Could have an Assume instead of an assert, but better to error out than silently cover it up if mistakes are introduced elsewhere I think.

Reply:
Do you think it'd be useful to add a new macro `AssertAndRecover()`:

- in debug builds, this asserts (terminates) if the condition fails
- in release builds, this executes some recovery code and continues.

eg here the recovery code would be to clear the `vAddrToSend` vector and return from `MaybeSendAddr()`. Better to not relay addr records to a peer and continue operation than crash the node.

Reply:
Isn't `AssertAndRecover(condition, recovery)` the same as `if (!Assume(condition)) { recovery; }` ? (Err, at least if `Assume` was the identity function its comments claim it to be, rather than being cast to `(void)` in release builds)

Not sure the recovery code is needed here -- more than MAX_ADDR_TO_SEND would be a protocol violation, but shouldn't cause a crash or anything, I think?

Reply:
> Isn't AssertAndRecover(condition, recovery) the same as if (!Assume(condition)) { recovery; }

Ah interesting. I didn't realise that.

> Not sure the recovery code is needed here -- more than MAX_ADDR_TO_SEND would be a protocol violation, but shouldn't cause a crash or anything, I think?

I think it's probably better to skip sending the `addr` message than violate the p2p protocol.

Reply:
#21317 has now been merged so I've updated this to use the `if (!Assume(condition)) { recovery; }` pattern.